sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruning.scala

```diff
@@ -110,7 +110,17 @@ private[sql] object ParquetSchemaPruning extends Rule[LogicalPlan] {
     val projectionRootFields = projects.flatMap(getRootFields)
     val filterRootFields = filters.flatMap(getRootFields)
 
-    (projectionRootFields ++ filterRootFields).distinct
+    // Kind of expressions don't need to access any fields of a root fields, e.g., `IsNotNull`.
+    // For them, if there are any nested fields accessed in the query, we don't need to add root
+    // field access of above expressions.
+    // For example, for a query `SELECT name.first FROM contacts WHERE name IS NOT NULL`,
+    // we don't need to read nested fields of `name` struct other than `first` field.
```

Comment:
I'm having trouble accepting this, but perhaps I'm reading too much into it (or not enough). Let me illustrate with a couple of queries and their physical plans.

Assuming the data model in `ParquetSchemaPruningSuite.scala`, the physical plan for the query

    select employer.id from contacts where employer is not null

is

```
== Physical Plan ==
*(1) Project [employer#36.id AS id#46]
+- *(1) Filter isnotnull(employer#36)
   +- *(1) FileScan parquet [employer#36,p#37] Batched: false, Format: Parquet,
    PartitionCount: 2, PartitionFilters: [], PushedFilters: [IsNotNull(employer)],
    ReadSchema: struct<employer:struct<id:int>>
```

The physical plan for the query

    select employer.id from contacts where employer.id is not null

is

```
== Physical Plan ==
*(1) Project [employer#36.id AS id#47]
+- *(1) Filter (isnotnull(employer#36) && isnotnull(employer#36.id))
   +- *(1) FileScan parquet [employer#36,p#37] Batched: false, Format: Parquet,
    PartitionCount: 2, PartitionFilters: [], PushedFilters: [IsNotNull(employer)],
    ReadSchema: struct<employer:struct<id:int>>
```

The read schemata are the same, but the query filters are not. The file scan for the second query looks as I would expect, but the scan for the first query appears to only read `employer.id` even though it needs to check `employer is not null`. If it only reads `employer.id`, how does it check that `employer.company` is not null? Perhaps `employer.id` is null but `employer.company` is not null for some row...

I have run some tests to validate that this PR is returning the correct results for both queries, and it is. But I don't understand why.

Reply:
For the first query, the constrain is `employer is not null`. 

When `employer.id` is not `null`, `employer` will always  not be `null`; as a result, this PR will work. 

However, when `employer.id` is `null`, `employer` can be `null` or `something`, so we need to check if `employer` is `something` to return a null of `employer.id`.

I checked in the `ParquetFilter`, `IsNotNull(employer)` will be ignored since it's not a valid parquet filter as parquet doesn't support pushdown on the struct; thus, with this PR, this query will return wrong answer. 

I think in this scenario, as @mallman suggested, we might need to read the full data.

Reply:
> I checked in the ParquetFilter, IsNotNull(employer) will be ignored since it's not a valid parquet filter as parquet doesn't support pushdown on the struct; thus, with this PR, this query will return wrong answer.

We may not worry about wrong answer from datasource like Parquet in predicate pushdown. As not all predicates are supported by pushdown, we always have a SparkSQL Filter on top of scan node to make sure to receive correct answer.

Reply:
@viirya, I see your point about the difference between a complex type being null and a subfield being null. So to answer the following query

    select address from contacts where name is not null

do we need to read any of the fields in `name`? Or perhaps just read one arbitrary field of simple type, like `name.first`? That's surprising, but I'm starting to believe it's true.

Reply:
Currently under this PR, `name` will be fully read. This is not perfect. However, to pick one arbitrary field from `name` sounds a little bit hacky to me. WDYT? cc @dbtsai @cloud-fan

Reply:
Btw, I think this is not seen as schema pruning case in the sense of original PR, so maybe we can leave it as it for now.

Reply:
Yeah, I'm okay with leaving it as-is.

Reply:
Instead of reading any arbitrary field of simple type (which may not exist if it's a deeply nested struct), I think we should implement the pushdown with complex type in parquet with similar logic, and let parquet reader handle it. 

@viirya Can you create a followup JIRA for this? 

Thanks.